consensus/src/consensus/head_requests.rs

```diff
@@ -151,9 +169,28 @@ impl<TNetwork: Network + 'static> Future for HeadRequests<TNetwork> {
                 (Ok(Ok(block)), peer_id) => {
                     self.unknown_blocks.push((block, peer_id));
                 }
-                // We don't do anything with failed requests.
                 (Ok(Err(error)), peer_id) => {
-                    trace!(%error, %peer_id, "Block request failed on remote side");
+                    match error {
+                        // We don't do anything with benign failed requests.
+                        BlockError::Other | BlockError::TargetHashNotFound => {
+                            trace!(%error, %peer_id, "Block request failed on remote side")
+                        }
+                        // We disconnect from the malicious peer in case of deliberate lying
+                        BlockError::ResponseHashMismatch => {
+                            warn!(%peer_id,
+                                "Banning peer due to wrong reply to requested block",
+                            );
+
+                            let network = Arc::clone(&self.network);
+                            spawn(Box::pin({
+                                async move {
+                                    network
+                                        .disconnect_peer(peer_id, CloseReason::MaliciousPeer)
+                                        .await;
+                                }
+                            }));
+                        }
+                    }
```

Comment:
I prefer this for it uses less intendation:
```rust
// We disconnect from the malicious peer in case of deliberate lying
(Ok(Err(BlockError::ResponseHashMismatch)), peer_id) => {
    warn!(%peer_id,
        "Banning peer due to wrong reply to requested block",
    );

    let network = Arc::clone(&self.network);
    spawn(Box::pin({
        async move {
            network
                .disconnect_peer(peer_id, CloseReason::MaliciousPeer)
                .await;
        }
    }));
}
// We don't do anything with failed requests.
(Ok(Err(error)), peer_id) => {
  trace!(%error, %peer_id, "Block request failed on remote side")
}
```